File: Problems/108_Convert_Sorted_Array_to_Binary_Search_Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# cleaner solution
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        
        def build(left, right):
            if left > right:
                return None
            middle = (left+right) // 2
            father = TreeNode(nums[middle])
            father.left  = build(left, middle-1)
            father.right = build(middle+1, right)
            return father
        return build(0, len(nums)-1)

class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        
        def toBST(node, nums_left, nums_right):
            if len(nums_left)  == 0:
                return
            node.left  = TreeNode(nums_left[len(nums_left)//2])
            toBST(node.left, nums_left[:len(nums_left)//2], nums_left[len(nums_left)//2+1:])
            if len(nums_right) == 0:
                return
            node.right = TreeNode(nums_right[len(nums_right)//2])
            toBST(node.right, nums_right[:len(nums_right)//2], nums_right[len(nums_right)//2+1:])
        ans = TreeNode(nums[len(nums)//2])
        toBST(ans, nums[:len(nums)//2], nums[len(nums)//2+1:])

        return ans
        

File: Problems/108_Convert_Sorted_Array_to_Binary_Search_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return build(nums, 0, nums.size()-1);
    }
    TreeNode* build(vector<int>& nums, int left, int right) {
        if (left > right) return nullptr;
        int middle = (left+right) / 2;
        TreeNode* father = new TreeNode(nums[middle]);
        father -> left = build(nums, left, middle-1);
        father -> right = build(nums, middle+1, right);
        return father;
    }
};

class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return convert(nums, 0, nums.size()-1);
    }
    
    TreeNode* convert(vector<int>& nums, int left, int right){
        if(left==right){
            return new TreeNode(nums[left]);
        }
        if(left>right){
            return nullptr;
        }
        int middle = (left+right)/2;
        TreeNode* node = new TreeNode(nums[middle]);
        node->left = convert(nums, left, middle-1);
        node->right = convert(nums, middle+1, right);
        return node;
    }
};